Add support for port names and labels

gkleiman opened the pull request.

aquamatthias commented on docs/docs/rest-api/public/api/v2/schema/AppDefinition.json:
```diff
@@ -131,6 +131,17 @@
+                    "description": "Name of the service hosted on this port."
+                  },
                   "protocol": {
                     "type": "string",
```
Not introduced by you: but we allow only tcp and udp.

gkleiman replied:
:+1:

aquamatthias commented on docs/docs/rest-api/public/api/v2/schema/AppDefinition.json:
```diff
@@ -405,6 +416,37 @@
+            "description": "Name of the service hosted on this port."
+          },
+          "protocol": {
+            "type": "string",
```
We allow only tcp and udp.

gkleiman replied:
:+1:

aquamatthias commented on docs/docs/rest-api/public/api/v2/schema/AppDefinition.json:
```diff
@@ -405,6 +416,37 @@
+          },
+          "name": {
+            "type": "string",
+            "description": "Name of the service hosted on this port."
```
Please add: if provided, it must be unique over all ports.

gkleiman replied:
:+1:

aquamatthias commented on src/main/proto/mesos/mesos.proto:
```diff
@@ -1,20 +1,18 @@
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
+// Licensed to the Apache Software Foundation (ASF) under one
```
We copy this proto from time to time - it is hard to see, which version this is copied from.
Please add a comment about that in the first line (and also were the original lives)

gkleiman replied:
:+1:

aquamatthias commented on src/main/scala/mesosphere/marathon/api/v2/json/AppUpdate.scala:
```diff
@@ -132,7 +125,8 @@ object AppUpdate {
     appUp.upgradeStrategy is valid
     appUp.storeUrls is optional(every(urlCanBeResolvedValidator))
-    appUp.ports is optional(elementsAreUnique(AppDefinition.filterOutRandomPorts))
+    appUp.ports is optional(elementsAreUnique(AppDefinition.filterOutRandomPorts, "Ports must be unique."))
```
Please remove `ports` and use `portDefinitions`

gkleiman replied:
:+1:

aquamatthias commented on src/main/scala/mesosphere/marathon/api/v2/json/AppUpdate.scala:
```diff
@@ -48,6 +39,8 @@ case class AppUpdate(
 
     ports: Option[Seq[Int]] = None,
```
Please remove `ports` and use `portDefinitions`

gkleiman replied:
:+1:

aquamatthias commented on src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala:
```diff
@@ -581,28 +585,41 @@ trait AppAndGroupFormats {
+              appWithoutPorts || extra.ipAddress.isEmpty
+            }
+            .filter(ValidationError("You cannot specify both ports and port definitions")) { extra =>
+              extra.maybePorts.isEmpty || extra.maybePortDefinitions.isEmpty
```
As you already suggested: if ports and portDefinitions hold the same data, we could allow this.

gkleiman replied:
:+1:

aquamatthias commented on src/main/scala/mesosphere/marathon/state/AppDefinition.scala:
```diff
@@ -82,7 +82,7 @@ case class AppDefinition(
   import mesosphere.mesos.protos.Implicits._
 
-  require(ipAddress.isEmpty || ports.isEmpty, "IP address and ports are not allowed at the same time")
+  require(ipAddress.isEmpty || portDefinitions.isEmpty, "IP address and ports are not allowed at the same time")
```
Do we have this also as part of the validation?

gkleiman replied:
We have the following validation in Formats.scala:

```scala
.filter(ValidationError("You cannot specify both an IP address and ports")) { extra =>
  val appWithoutPorts = extra.maybePorts.forall(_.isEmpty) && extra.maybePortDefinitions.forall(_.isEmpty)
  appWithoutPorts || extra.ipAddress.isEmpty
}
```

aquamatthias commented on src/main/scala/mesosphere/marathon/state/AppDefinition.scala:
```diff
@@ -248,12 +256,12 @@ case class AppDefinition(
 
   def hostPorts: Seq[Int] =
-    containerHostPorts.getOrElse(ports.map(_.toInt))
+    containerHostPorts.getOrElse(portDefinitions.map(_.port))
```
getOrElse(ports)

gkleiman replied:
:+1: renamed AppDefinition.port to AppDefinition.portNumbers and made it a lazy val.

aquamatthias commented on src/main/scala/mesosphere/marathon/state/AppDefinition.scala:
```diff
@@ -248,12 +256,12 @@ case class AppDefinition(
 
   def servicePorts: Seq[Int] =
-    containerServicePorts.getOrElse(ports.map(_.toInt))
+    containerServicePorts.getOrElse(portDefinitions.map(_.port))
```
getOrElse(ports)

gkleiman replied:
:+1:

aquamatthias commented on src/main/scala/mesosphere/marathon/state/PortDefinition.scala:
```diff
@@ -0,0 +1,48 @@
+    }
+  }
+
+  implicit val portDefinitionsValidator: Validator[Seq[PortDefinition]] = validator[Seq[PortDefinition]] {
```
Please rename to something more expressive: e.g. `portsAreUnique`

gkleiman replied:
We validate more than that:
 - Port numbers must be unique,
 - Port names must be unique.
 - Port numbers must be >= 0

aquamatthias replied:
Suggestion: `portDefinitionsValid`

aquamatthias commented on src/main/scala/mesosphere/marathon/api/v2/json/AppUpdate.scala:
```diff
@@ -132,7 +125,8 @@ object AppUpdate {
     appUp.storeUrls is optional(every(urlCanBeResolvedValidator))
-    appUp.ports is optional(elementsAreUnique(AppDefinition.filterOutRandomPorts))
+    appUp.ports is optional(elementsAreUnique(AppDefinition.filterOutRandomPorts, "Ports must be unique."))
+    appUp.portDefinitions is optional(PortDefinitions.portDefinitionsValidator)
```
If you import PortDefinitions and rename the validator, it could look like:
`appUp.portDefinitions is optional(portsAreUnique)`

aquamatthias commented on src/main/scala/mesosphere/marathon/state/AppDefinition.scala:
```diff
@@ -469,7 +478,7 @@ object AppDefinition {
     appDef.container.each is valid
     appDef.storeUrls is every(urlCanBeResolvedValidator)
-    appDef.ports is elementsAreUnique(filterOutRandomPorts)
+    appDef.portDefinitions is PortDefinitions.portDefinitionsValidator
```
See my other comment - please rename.

aquamatthias commented on src/main/scala/mesosphere/marathon/state/GroupManager.scala:
```diff
@@ -203,7 +201,8 @@ class GroupManager @Singleton @Inject() (
   private[state] def assignDynamicServicePorts(from: Group, to: Group): Group = {
     val portRange = Range(config.localPortMin(), config.localPortMax())
-    var taken = from.transitiveApps.flatMap(_.ports) ++ to.transitiveApps.flatMap(_.ports)
+    var taken = from.transitiveApps.flatMap(_.portDefinitions.map(_.port)) ++
```
why can't we use the original statement? ports should be fine.

gkleiman replied:
:+1:

aquamatthias commented on src/main/scala/mesosphere/marathon/state/GroupManager.scala:
```diff
@@ -216,12 +215,19 @@ class GroupManager @Singleton @Inject() (
       val assignedAndAvailable = mutable.Queue(
         from.app(app.id)
-          .map(_.ports.filter(p => portRange.contains(p) && !app.servicePorts.contains(p)))
+          .map(_.portDefinitions.map(_.port).filter(p => portRange.contains(p) && !app.servicePorts.contains(p)))
```
use `ports`

gkleiman replied:
:+1:

aquamatthias commented on src/main/scala/mesosphere/marathon/state/Migration.scala:
```diff
@@ -300,6 +305,56 @@ class MigrationTo0_13(taskRepository: TaskRepository, store: PersistentStore) {
+
+    for {
+      rootGroup <- rootGroupFuture
+      apps = rootGroup.transitiveApps
```
not needed?

gkleiman replied:
Shouldn't `rootGroup.transitiveApps` and `appRepository.apps()` return the same apps?

aquamatthias replied:
Should. ignore my comment.

aquamatthias commented on src/main/scala/mesosphere/marathon/state/PortDefinition.scala:
```diff
@@ -0,0 +1,48 @@
+    }
+  }
+
+  private[this] def uniquePortNames: Validator[Seq[PortDefinition]] = new Validator[Seq[PortDefinition]] {
```
Can you just use `Validator.elementsAreUnique`?

gkleiman replied:
`Validator.elementsAreUnique` only accepts a "filter" function that filter/maps from Seq[A] to Seq[A].

In this case we need to map from `Seq[PortDefinitions]` to `Seq[Int]` and then apply the validation logic.

That means that I can not use `Validator.elementsAreUnique` as-is. I'll try to refactor it later.

aquamatthias replied:
Suggestion: 
Validation.scala:
```scala
  def elementsAreUniqueBy[A, B](fn: A => B,
                              errorMessage: String = "Elements must be unique."): Validator[Seq[A]] = {
    new Validator[Seq[A]] {
      def apply(seq: Seq[A]) = {
        val filteredSeq = seq.map(fn)
        if (filteredSeq.size == filteredSeq.distinct.size) Success
        else Failure(Set(RuleViolation(seq, errorMessage, None)))
      }
    }
  }
```

PortDefinition.scala

```scala
      portDefinitions is valid(elementsAreUniqueBy(_.name))
```

aquamatthias commented on src/main/scala/mesosphere/marathon/state/PortDefinition.scala:
```diff
@@ -0,0 +1,48 @@
+  }
+
+  implicit val portDefinitionsValidator: Validator[Seq[PortDefinition]] = validator[Seq[PortDefinition]] {
+    portDefinitions: Seq[PortDefinition] => portDefinitions is every(valid) and uniquePorts and uniquePortNames
```
just a taste question: do we need this type attribute?

gkleiman replied:
Nope, removed it.

aquamatthias commented on src/main/scala/mesosphere/mesos/TaskBuilder.scala:
```diff
@@ -174,23 +173,23 @@ class TaskBuilder(app: AppDefinition,
-    app.ipAddress match {
-      case Some(IpAddress(_, _, DiscoveryInfo(ports))) if ports.nonEmpty =>
-        val builder = org.apache.mesos.Protos.DiscoveryInfo.newBuilder
+  protected def computeDiscoveryInfo(app: AppDefinition): org.apache.mesos.Protos.DiscoveryInfo = {
```
Do we have a test case for this?

gkleiman replied:
I added quite some test cases in `TaskBuilderTest.scala`.

aquamatthias replied:
:+1:

aquamatthias commented on src/test/scala/mesosphere/marathon/api/v2/json/AppDefinitionFormatsTest.scala:
```diff
@@ -60,7 +60,8 @@ class AppDefinitionFormatsTest
     (r1 \ "fetch").as[Seq[FetchUri]] should equal (DefaultFetch)
     (r1 \ "storeUrls").as[Seq[String]] should equal (DefaultStoreUrls)
-    (r1 \ "ports").as[Seq[Long]] should equal (DefaultPorts.map(_.toInt))
+    (r1 \ "ports").as[Seq[Long]] should equal (DefaultPortDefinitions.map(_.port))
```
remove ports.

gkleiman replied:
:+1:

aquamatthias commented on src/test/scala/mesosphere/marathon/api/v2/json/AppDefinitionFormatsTest.scala:
```diff
@@ -107,7 +108,8 @@ class AppDefinitionFormatsTest
     r1.fetch should equal (DefaultFetch)
     r1.storeUrls should equal (DefaultStoreUrls)
-    r1.ports should equal (DefaultPorts)
+    r1.ports should equal (DefaultPortDefinitions.map(_.port))
```
Remove ports.

gkleiman replied:
:+1:

aquamatthias commented on src/test/scala/mesosphere/marathon/api/v2/json/AppDefinitionTest.scala:
```diff
@@ -256,7 +330,7 @@ class AppDefinitionTest extends MarathonSpec with Matchers {
 
   test("SerializationRoundtrip empty") {
-    import Formats._
+    import TestFormats._
```
Hope we can get rid of those imports.

gkleiman replied:
We can :dancers:

aquamatthias commented on src/test/scala/mesosphere/marathon/api/v2/json/AppUpdateTest.scala:
```diff
@@ -43,13 +43,37 @@ class AppUpdateTest extends MarathonSpec {
+    shouldViolate(
+      update.copy(portDefinitions = Some(Seq(
+        PortDefinition(port = 9000, name = Some("foo")),
+        PortDefinition(port = 9001, name = Some("foo"))))
```
:+1:

aquamatthias commented on src/test/scala/mesosphere/marathon/state/MigrationTo0_16Test.scala:
```diff
@@ -0,0 +1,126 @@
+  }
+
+  test("an app and all its revisions are migrated") {
+    // FIXME (gkleiman): this test is flaky
```
Did you test migrating from an older version? 
And yes: please deflake :)

aquamatthias commented on src/test/scala/mesosphere/mesos/TaskBuilderTest.scala:
```diff
@@ -329,6 +373,9 @@ class TaskBuilderTest extends MarathonSpec with Matchers {
     TextFormat.shortDebugString(networkInfos.head) should equal(TextFormat.shortDebugString(networkInfoProto))
     networkInfos.head should equal(networkInfoProto)
+
+    taskInfo.hasDiscovery should be (true)
```
This should be true for all task protos that we create - correct?

gkleiman replied:
Yes, that's why I added the following to `assertTaskInfo`, in order to check this in existing tests that use that method:

```scala
    assert(taskInfo.hasDiscovery)
    val discoveryInfo = taskInfo.getDiscovery
    val discoveryInfoProto = MesosProtos.DiscoveryInfo.newBuilder
      .setVisibility(MesosProtos.DiscoveryInfo.Visibility.FRAMEWORK)
      .setName(taskInfo.getName)
      .setPorts(
        MesosProtos.Ports.newBuilder
          .addPorts(MesosProtos.Port.newBuilder.setNumber(8080).setProtocol("tcp"))
          .addPorts(MesosProtos.Port.newBuilder.setNumber(8081).setProtocol("tcp"))
      ).build
    TextFormat.shortDebugString(discoveryInfo) should equal(TextFormat.shortDebugString(discoveryInfoProto))
    discoveryInfo should equal(discoveryInfoProto)
```

aquamatthias commented on src/test/scala/mesosphere/mesos/TaskBuilderTest.scala:
```diff
@@ -329,6 +373,9 @@ class TaskBuilderTest extends MarathonSpec with Matchers {
     networkInfos.head should equal(networkInfoProto)
+
+    taskInfo.hasDiscovery should be (true)
+    taskInfo.getDiscovery.getName should be (taskInfo.getName)
```
Don't know if I missed that, but we should have at least one test case, that tests all properties of a discovery info.

gkleiman replied:
:+1:

aquamatthias commented:
@gkleiman Thanks for this addition - looks great!
I like your proposal for handling ports/portDefinitions in the json reads function.
Made a couple of comments. Please address and create a separate commit to review.

gkleiman commented:
@aquamatthias thanks for the review!

I addressed most of your comments in separate commits.

Pending tasks:
- [ ] Test the migration manually.
- [ ] "Deflake" the migration test (or fix the migration if it is broken).
- [ ] Write some documentation and a change log entry.
- [ ] Get rid of all of this and fix our ports API :sunglasses:

aquamatthias commented:
@gkleiman Looks good to me. Made 2 very minor suggestions (can be addressed in 5 mins).
- The deflake point is holding me back from merge.
- The documentation can be in a separate PR

gkleiman commented:
@aquamatthias squashed

aquamatthias commented:
Thanks! A big ship for that!
```



                              __    __    __
                             |==|  |==|  |==|
                           __|__|__|__|__|__|_
                        __|___________________|___
                     __|__[]__[]__[]__[]__[]__[]__|___
                    |............................o.../
                    \.............................../
                  _,~')_,~')_,~')_,~')_,~')_,~')_,~')/,~')_




```
